stdlib/src/memory/arc.mojo

```diff
@@ -45,32 +45,231 @@ print(Arc(String("ok"))[])
 """
 
 from os.atomic import Atomic
+from os import abort
 
 from builtin.builtin_list import _lit_mut_cast
 from memory import UnsafePointer, stack_allocation
+from memory.maybe_uninitialized import UnsafeMaybeUninitialized
+from collections import Optional, OptionalReg
 
 
-struct _ArcInner[T: Movable]:
+struct _ArcInner[T: Movable, enable_weak: Bool = False]:
     var refcount: Atomic[DType.uint64]
-    var payload: T
+    var payload: UnsafeMaybeUninitialized[T]
+    
+    alias _weak_refcount_type = __mlir_type[
+        `!pop.array<`, Int(enable_weak).value, `, `, Atomic[DType.int64], `>`
+    ]
+
+    var weak_refcount: Self._weak_refcount_type
 
     fn __init__(inout self, owned value: T):
         """Create an initialized instance of this with a refcount of 1."""
         self.refcount = 1
-        self.payload = value^
+        self.payload = UnsafeMaybeUninitialized(value^)
+        self.weak_refcount = __mlir_op.`kgen.undef`[_type = Self._weak_refcount_type]()
+        
+        @parameter
+        if enable_weak:
+            var p = self._weakref_ptr()
+            p.init_pointee_explicit_copy(Atomic[DType.int64](1))
+
+        
+    fn _weakref_ptr(inout self) -> UnsafePointer[Atomic[DType.int64]]:
+        constrained[enable_weak, "can't get ptr, since there is no weakref to point to"]()
+
+        @parameter
+        if enable_weak:
+            var ptr = __mlir_op.`pop.array.gep`(
+                UnsafePointer.address_of(self.weak_refcount).address,
+                Int(0).value
+            )
+        
+            return UnsafePointer(ptr)
+        else:
+            constrained[False, "something is horribly broken, and _weakref_ptr is being called in an Arc with enable_weak = False"]()
+            return UnsafePointer[Atomic[DType.int64]]()
+
+    fn _destroy_value(inout self):
+        """
+        Run the destructor for the payload,
+        with the assumption that it has not been run yet
+        """
+        self.payload.assume_initialized_destroy()
+
+    fn try_upgrade_weak(inout self) -> Bool:
```

Comment:
**Suggestion** I think this could use the "custom self types" / "conditional conformance" trick to be statically constrained to:

```suggestion
fn try_upgrade_weak(inout self: Arc[T, True]) -> Bool
```

and then the `@parameter if enable_weak` statement below could be simplified.